### PR TITLE
CTW-333: Filter out non-builder records using the builder ID in the JWT

### PR DESCRIPTION
**Pull request by melocule**
Components that were meant to display only records owned by the builder now address the case of shared records between builders. Rather than filtering out by non-builder tags, the query filters for the ownership tag of the specific builder.

### CODE REVIEW

File: src/fhir/search-helpers.ts [hunk 1]
```diff
@@ -60,17 +63,17 @@ export async function searchAllRecords<T extends ResourceTypeString>(
...
   searchParams?: SearchParams
 ): Promise<SearchReturn<T>> {
-  const nonBuilderTags = [...THIRD_PARTY_TAGS, ...LENS_TAGS];
-
+  const jwt = getJWT(fhirClient);
+  const builderTag = `${SYSTEM_ZUS_OWNER}|builder/${jwt[SYSTEM_ZUS_BUILDER_ID]}`;
   return searchAllRecords(resourceType, fhirClient, {
     ...searchParams,
-    "_tag:not": nonBuilderTags.join(","),
+    _tag: builderTag,
```
Comment (luthfib) [inline src/fhir/search-helpers.ts:76]: This doesnt look right to me. Might have to look around and ensure but I think you would create duplicates with this. You might be able to do something like: 

```
  _tag:not": nonBuilderTags.join(","),
  _tag: builderTag,
```

**Review by luthfib — CHANGES_REQUESTED** [review]
Looking good but a couple of show stoppers

**Comment by luthfib** [comment]
I also wonder if we want to get the auth from the FHIR client or if we want to allow the ctw provider to pass the auth token, as I think that will be a more generic approach that will decouple the FHIR client and the auth token.. thoughts @WesleyKapow

File: src/fhir/client.ts [hunk 2]
```diff
@@ -8,13 +9,33 @@ export function getFhirClient(env: Env, accessToken: string) {
...
+
+  return client;
+}
+
+export type ZusJWT = {
+  SYSTEM_ZUS_BUILDER_ID: string;
+  // We only use the above properties. Others can be any key with any type.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: string]: any;
+};
```
[the remark this answers is not in the document]
Comment (melocule) [reply src/fhir/client.ts:28]: There's a property in the JWT that's actually an array of strings, just realized I can type it as `string | string[]` though so I'm going with that!

File: src/fhir/search-helpers.ts (hunk 1 above)
Comment (melocule) [reply src/fhir/search-helpers.ts:76]: Thanks for the catch! Didn't see any resources with both, but apparently whether those tags overlaps can change with time.

**Comment by WesleyKapow** [comment]
@luthfib agreed that getting it from FHIRClient is hacky. I'm just not thrilled with having to pass the accessToken around everywhere which is why I was opting for the hacky approach for now (no refactoring or plumbing of an argument needed).

Eventually we may need the `CTWProvider` to hand us something that we can pass around everywhere and hang stuff off of that. Things like `env`, `accessToken`, user details, etc.

File: src/fhir/client.ts [hunk 3]
```diff
@@ -17,6 +18,28 @@ export function getFhirClient(env: Env, accessToken: string) {
...
 }
 
+export type ClientAuth = {
+  httpClient: {
+    authHeader: { authorization: string };
+  };
+} & Client;
+
+export type ZusJWT = {
+  SYSTEM_ZUS_BUILDER_ID: string;
```
Comment (WesleyKapow) [inline src/fhir/client.ts:28]: Here you are saying this has a key of the literal string: `SYSTEM_ZUS_BUILDER_ID`.

File: src/fhir/client.ts [hunk 4]
```diff
@@ -17,6 +18,28 @@ export function getFhirClient(env: Env, accessToken: string) {
...
+  // We only use the above properties. Others can be any key with any type.
+  [key: string]: string | string[];
+};
+
+export function getJWT(fhirClient: Client): ZusJWT {
```
Comment (WesleyKapow) [inline src/fhir/client.ts:33]: s/getJWT/getClaims

The accessToken is the JWT, I believe you are decoding it and getting the claims data.

**Review by WesleyKapow — CHANGES_REQUESTED** [review]

File: src/fhir/client.ts [hunk 5]
```diff
@@ -17,6 +18,28 @@ export function getFhirClient(env: Env, accessToken: string) {
   });
 }
 
+export type ClientAuth = {
```
Comment (WesleyKapow) [inline src/fhir/client.ts:21]: Need to be exported?

**Review by WesleyKapow — APPROVED** [review]
Side note -- avoid super-long-branch-names-that-describe-everything :P (it could have been something like `CTW-333/filter-by-owner-tag`)

**Comment by WesleyKapow** [comment]
@melocule may need to resolve merge conflicts with main to get CI going.. not sure.. CI has been a little weird lately.

**Review by luthfib — APPROVED** [review]

**Pull request merged**